### PR TITLE
Clear Publish Date Button

### LINKED ISSUE #12048: No way to clear publish date from datepicker once set
Once you set a publish date using the publish datepicker, there is no way to go back and remove that if you would like to have it publish immediately on press of the publish button. The best you can do is just manually set it in the past or close to the current time. 

In the classic editor, there was a cancel button that would clear the scheduled date. 

publish-immediately

Similar: #9030

### PR DESCRIPTION
**Pull request by earnjam**
## Description
This PR adds an `Unschedule` button to the sidebar below the Publish Date. The button only shows when: 
- A publish date has been set
- The post is not published
- The post is not scheduled

It allows the user to clear the selected publish date and returns it to a floating date, displaying 'Immediately'. 

This is the simplest implementation of the [mock-ups from the ticket](https://github.com/WordPress/gutenberg/issues/12048#issuecomment-510052209) by @karmatosed . 

Fixes #12048

## Screenshots 
Kapture 2020-03-15 at 21 14 52

## Types of changes
New feature (non-breaking change which adds functionality) 

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by earnjam** [comment]
The failing test is just related to `clone-deep` changing the resolved NPM registry URL to `https` instead of `http`. Just need someone to update package-lock.json to fix that. Otherwise this passes.

File: packages/editor/src/components/post-schedule/index.js [hunk 1]
```diff
@@ -6,7 +6,7 @@ import { withSelect, withDispatch } from '@wordpress/data';
 import { compose } from '@wordpress/compose';
 import { DateTimePicker } from '@wordpress/components';
 
-export function PostSchedule( { date, onUpdateDate } ) {
+export function PostSchedule( { date, modified, onUpdateDate } ) {
```
Comment (earnjam) [inline packages/editor/src/components/post-schedule/index.js]: One option here instead of adding new props to `<PostSchedule>` would be to simply use the current time down below. We just need to pass something other than `null` to `<DateTimePicker>` 
 in `currentDate` or else it acts up when the date has been cleared and you try to set another one. It immediately closes and uses the current time no matter what you click inside it.

**Comment by karmatosed** [comment]
I think from a design perspective this should be a link not a button. That would visually make more sense right now.

**Comment by earnjam** [comment]
> I think from a design perspective this should be a link not a button. That would visually make more sense right now. I realise a while ago a button was recommended but we're going back to using links for a lot of things, so this feels right.

That would definitely take some of the weight of it away. The button does feel a bit heavy for the space at the moment. I'll get the PR updated with a link.

File: packages/editor/src/components/post-schedule/index.js (hunk 1 above)
Comment (earnjam) [reply packages/editor/src/components/post-schedule/index.js]: I got rid of this and moved to using current time instead

**Comment by earnjam** [comment]
Switched it over to a link, but there is still a bug that needs to be worked out. 

Focus is set on the day field in the datepicker when it is opened, and then the date gets updated `onBlur` from the value in that field. This means that if you have cleared the date, then it gets set to the current day and closes the datepicker when you try to page ahead to another month.

File: packages/editor/src/components/post-schedule/index.js (hunk 1 above)
Comment (earnjam) [reply packages/editor/src/components/post-schedule/index.js]: Nevermind, has to be the `modified` time because it will get set `onBlur` and we want it to remain to be considered floating until it's specifically set to something else.

**Comment by mapk** [comment]
Looking good, @earnjam. I've got just a few design changes.

**Current iteration in PR:**

Screen Shot 2020-03-18 at 11 53 23 AM

**Proposed design improvements:**

- Remove the `is-small` class. That size looks awkward right now.
- Move the "unschedule" link closer to the scheduled time. This way the design principle of Proximity helps the user make a connection.

Screen Shot 2020-03-18 at 11 57 31 AM

**Comment by earnjam** [comment]
Thanks @mapk! Updated with the suggested changes. 

image

**Comment by earnjam** [comment]
Added an e2e test

**Review by mapk — APPROVED** [review]
Looks good from a design perspective and worked as expected! Thanks!

Screen Shot 2020-03-31 at 9 56 03 AM

**Comment by mtias** [comment]
I think this action should be added to the calendar popover itself so the flow of setting / unsetting a date is more consistent.

**Comment by earnjam** [comment]
> I think this action should be added to the calendar popover itself so the flow of setting / unsetting a date is more consistent.

I agree in principle, but if you read the conversation from #12048, there are a lot of a11y issues to consider for the calendar popover that make adding save/cancel buttons a complicated problem to solve. This was the short term solution proposed while a more robust design effort could be made there.

**Comment by mapk** [comment]
@youknowriad or @jorgefilipecosta can we get a technical review on this before merging?

File: packages/editor/src/components/post-schedule/index.js [hunk 2]
```diff
@@ -33,6 +33,9 @@ export default compose( [
 	withSelect( ( select ) => {
 		return {
 			date: select( 'core/editor' ).getEditedPostAttribute( 'date' ),
+			modified: select( 'core/editor' ).getEditedPostAttribute(
+				'modified'
```
Comment (youknowriad) [inline packages/editor/src/components/post-schedule/index.js:37]: what is this "modified" attribute? and why it's being used here? Also, it seems that this change is unrelated with the root issue of the PR (unschedule link)?

**Comment by mtias** [comment]
A lot of considerations from that ticket, for sure, but I still think a "reset" added to the date popover is more intuitive and helpful for the component's behaviour for others that might make use of it.

File: packages/editor/src/components/post-schedule/index.js (hunk 2 above)
Comment (earnjam) [reply packages/editor/src/components/post-schedule/index.js:37]: `modified` is the timestamp the post was last modified. `modified === date` is one of the ways we determine that the post is considered to have a floating date. 

If you pass a `null` value for `currentDate` to the `DateTimePicker` then it uses the current date to pre-populate the input fields. Then `onBlur` it will save the current value of the input as `date`, effectively eliminating the floating date we just set. It immediately closes and uses the current time no matter what you click inside it. 

The REST API expects `null` in order to make a date floating, so that's why we're using that for the date.

**Comment by karmatosed** [comment]
I have mocked up what that reset might look like here:

Frame 1

It's super simple but should work. I don't think we need it to be 'red' as it's not deleting just resetting.

**Comment by mapk** [comment]
I think Tammie's mockup works well. I would imagine that resetting a calendar date would reset it to the current time/date, so makes sense to me.

@earnjam, can this PR pivot a bit here? Let's remove the "unschedule" link and add a "reset" link to the calendar component. Is that workable for you?

**Comment by earnjam** [comment]
No problem! I'll work on moving it inside the datepicker popover and get this PR updated.

**Comment by earnjam** [comment]
I moved put a `Reset` button inside the `DateTimePicker` component as shown in Tammie's mockup. Here's a quick gif of how it works:
date-reset

I also rebased `master` just to make sure this doesn't fall too far behind.  

A few things to note here:
---
1) I'm still passing the post modified date into the `DateTimePicker` component as a fallback for when the date is `null`. The reason for this is a compound problem related to the functionality of the component. 

The `DatePicker` child component (calendar) accepts and uses a `null` value for the `currentDate` prop, but the `TimePicker` does not for its equivalent `currentTime` prop. It can't really since it has generic input fields, so it needs to pre-populate them with something. With a `null` value, it will just fall back to using the current time. The problem is that it also fires the `onChange` callback whenever the focus moves away from any field inside it, causing the date to save immediately. When the `DateTimePicker` opens in the popover, the day input of the `TimePicker` is focused. If you then click anywhere inside the popover (such as the arrow to go to the next month, or a day on the calendar), it will close and set the date for the post to the current time. 

So the only options I saw to solve this were:
- Redesign the component to have some sort of `Save` button that saves and closes the popover and stop saving on focus change
- Pass in the `modified` date as the fallback. This is sort of hacky, but works because it's also how we initialize the editor and is one of the criteria for considering a post date floating (`date === modified`)

The latter is a much easier approach for now, short of a bigger effort to improve component as a whole. 
---
2) Because of the issue above, the reset button has to be visible all the time. I can't hide it based off a `null` date prop because I can't pass one in. Any checks to the state would force the `DateTimePicker` to have a dependency on the editor, so that's not acceptable. 
---
3) When the reset button is clicked, it is clearing the focus to close the popover because there is not a way to trigger the `Dropdown` to close from inside the `renderContent` function (possible improvement that could be made there). Blowing away the focus is not good from an A11y perspective. It is, however, the same action we're taking for any click on a day in the `DatePicker`, so it's not a huge deviation. We could possibly send focus to the date toggle that opens the popover, but my quick test of that prevented the popover from closing. Might be able to find a way to make that work though. 
---
4) Finally, and this is minor, but I don't love `components-datetime__buttons` that I used for the class name for the `<div>` wrapping the buttons on the `DateTimePicker`, but couldn't think of anything better that was generic.

**Comment by mapk** [comment]
I tested this and was able to schedule a post, reset that publish date and then publish the post immediately. It works as expected. 

Let's get some technical feedback before merging. cc @youknowriad

**Comment by youknowriad** [comment]
Great work here

> also fires the onChange callback whenever the focus moves away from any field inside it, causing the date to save immediately

That's the only thing bothering me in this PR. I feel it's a bit of a hack, and I wonder if there's a small fix like: on focus out, if the value was "untouched" (a local boolean to keep track of when the value really changes from the one used as a fallback) trigger the onChange, otherwise don't.

cc @ItsJonQ in case you have an idea about how to represent the "current" time (null value) in  the TimePicker component properly.

**Comment by earnjam** [comment]
> I feel it's a bit of a hack, and I wonder if there's a small fix like: on focus out, if the value was "untouched" (a local boolean to keep track of when the value really changes from the one used as a fallback) trigger the onChange, otherwise don't.

Good call. I looked into `TimePicker` more and we already store the initialized/previous value in state for the component, so in the update functions we can simply compare to that and if it hasn't changed, bail early and don't save. I just pushed a commit that does that and pulled out the logic using `modified` to initialize the component. 

Seems to work great and also solves the issue of accidentally setting the date on a new post just by tabbing past the month field.

**Review by youknowriad — APPROVED** [review]
Nice work here

**Comment by mtias** [comment]
Thanks for this improvement!

**Comment by mapk** [comment]
Really great work! I'm so excited about this one. 🎉

**Comment by ItsJonQ** [comment]
>  cc @ItsJonQ in case you have an idea about how to represent the "current" time (null value) in the TimePicker component properly.

@youknowriad 

I've been noticing a trend of... "how do we handle value parsing + resetting for controls?" lately 😂 . Some from PRs, but mostly from my work on padding controls.

I'm starting to get a sense of a "standardized" pattern of doing things. However, applying it to the DatePicker. That I'm unsure of (for now)
